python_ta/checkers/possibly_undefined_checker.py

```diff
@@ -0,0 +1,157 @@
+
+
+class PossiblyUndefinedChecker(BaseChecker):
+
+    __implements__ = IAstroidChecker
+    # name is the same as file name but without _checker part
+    name = 'possibly_undefined'
+    # use dashes for connecting words in message symbol
+    msgs = {'E9969': ('This variable might not be defined when this statement is executed.',
+                      'possibly-undefined',
+                      'Reported when a statement uses a variable that might not be assigned.'),
+           }
+
+    # this is important so that your checker is executed before others
+    priority = -1
+
+    possibly_undefined = set()
+
+    @check_messages('possibly-undefined')
+    def visit_name(self, node):
+        if self._check_possibly_undefined(node):
+            self.add_message('possibly-undefined', node=node)
+
+    def visit_module(self, node):
+        self._analyze(node)
+
+    def visit_functiondef(self, node):
+        self._analyze(node)
+
+    def _check_possibly_undefined(self, node: astroid.Name) -> bool:
+        """Returns True if there exists a path from the start block to node where
+        a variable used by node might not be defined."""
+        return node in self.possibly_undefined
+
+    def _analyze(self, node: Union[astroid.Module, astroid.FunctionDef]) -> None:
+        """Runs the data flow algorithm on a `Module` or `Function` CFG, which in turn
+        appends `Name` nodes to `possibly_undefined` if it might not be defined.
+        """
+        facts = {}
+        blocks = self._get_blocks_po(node)
+
+        all_assigns = self._get_assigns(blocks)
+        for block in blocks:
+            facts[block] = {}
+            facts[block]['out'] = all_assigns.copy()
+
+        worklist = blocks[:]
+        while len(worklist) != 0:
+            b = worklist.pop()
+            outs = [facts[p.source]['out'] for p in b.predecessors if p.source in facts]
+            if outs == []:
+                facts[b]['in'] = set()
+            else:
+                facts[b]['in'] = set.intersection(*outs)
+            temp = self._transfer(b, facts[b]['in'], all_assigns)
+            if temp != facts[b]['out']:
+                facts[b]['out'] = temp
+                successors = set([succ.target for succ in b.successors])
+                worklist = list(set(worklist).union(successors))
+
+    def _transfer(self, block: CFGBlock, in_facts: Set[str], all_assigns: Set[str]) -> Set[str]:
+        gen = set()
+        kill = set()
+        for statement in block.statements:
+            if isinstance(statement, astroid.Assign):
+                gen.update(self._get_targets(statement))
+            elif isinstance(statement, astroid.AnnAssign) and hasattr(statement.target, 'name'):
+                gen.add(statement.target)
+            elif isinstance(statement, astroid.Delete):
+                kill.update(self._get_targets(statement))
+            self._analyze_statement(statement, gen.union(in_facts.difference(kill)), all_assigns)
```

Comment:
This part can be simplified; it shouldn't be necessary to do a full recursion, but instead do a `for name in statements.nodes_of_class(astroid.Name): ...`.